linux/bindings/src/crypto/aead.rs

```diff
@@ -0,0 +1,318 @@
+        unsafe {
+            to_result(crate::crypto_aead_setkey(
+                self.aead.as_ptr(),
+                key.as_ptr(),
+                key.len() as u32,
+            ))
+        }
+    }
+}
+
+#[pinned_drop]
+impl PinnedDrop for Aead {
+    fn drop(self: Pin<&mut Self>) {
+        // SAFETY: `self.aead` is constructed by `new`, it is non-null and valid.
+        unsafe {
+            let aead = self.aead.as_ptr();
+            let base = addr_of_mut!((*aead).base);
+            crate::crypto_destroy_tfm(aead as _, base);
+        }
+    }
+}
+
+/// An `Aead` request.
+///
+/// This is constructed by the `Aead`, and is used to construct a
+/// cipher operation then handle it.
+pub struct AeadReq {
+    req: NonNull<crate::aead_request>,
+}
+
+impl AeadReq {
+    /// The kernel crypto APIs use `scatterlist` to transfer memory buffer.
+    ///
+    /// The memory structure for AEAD cipher operation has the following structure:
+    ///
+    /// - AEAD encryption input:  assoc data || plaintext
+    /// - AEAD encryption output: assoc data || ciphertext || auth tag
+    /// - AEAD decryption input:  assoc data || ciphertext || auth tag
+    /// - AEAD decryption output: assoc data || plaintext
+    const SCATTERLIST_SIZE: usize = 3;
+
+    /// Sets information needed to perform the cipher operation.
+    ///
+    /// The `src` and `dst` hold the associated data concatenated with the
+    /// plaintext or ciphertext. And `assoclen` specifies the length of
+    /// associated data in `src`, `cryptlen` specifies the number
+    /// of bytes to process from `src`, `iv` specifies the IV for the cipher
+    /// operation which must comply with the IV size returned by `Aead::ivsize`.
+    fn set_crypt(&self, src: &SgTable, dst: &SgTable, assoclen: usize, cryptlen: usize, iv: &[u8]) {
+        // SAFETY: `self.req` is non-null and valid.
+        unsafe {
+            let req_src = addr_of_mut!((*self.req.as_ptr()).src);
+            *req_src = (*src.0.get()).sgl;
+            let req_dst = addr_of_mut!((*self.req.as_ptr()).dst);
+            *req_dst = (*dst.0.get()).sgl;
+            let req_assoclen = addr_of_mut!((*self.req.as_ptr()).assoclen);
+            *req_assoclen = assoclen as _;
+            let req_cryptlen = addr_of_mut!((*self.req.as_ptr()).cryptlen);
+            *req_cryptlen = cryptlen as _;
+            let req_iv = addr_of_mut!((*self.req.as_ptr()).iv);
+            *req_iv = iv.as_ptr() as _;
+        }
+    }
+
+    /// Handles encryption request.
+    ///
+    /// Returns `Ok(())` if the cipher operation was successful, or an error
+    /// code in [`Error`].
+    pub fn encrypt(
+        self,
+        aad: &[u8],
+        input: &[u8],
+        iv: &[u8],
+        output: &mut [u8],
+        mac: &mut [u8],
+    ) -> Result {
+        let mut src = SgTable::alloc(Self::SCATTERLIST_SIZE)?;
+        let mut dst = SgTable::alloc(Self::SCATTERLIST_SIZE)?;
```

Comment:
Consider the error handling part here: If `dst` failed in allocation while `src` succeeded, would the corresponding SgTable's memory of `src` be freed properly?

Reply:
When `src` is dropped, `sg_free_table` will be called to free the memory.